File: frontend/leetlab/lib/model/problem.dart
import 'package:leetlab/model/constraint.dart';
import 'package:leetlab/model/example.dart';
import 'package:leetlab/model/hint.dart';
import 'package:leetlab/model/starter_code.dart';
import 'package:leetlab/model/test_case.dart';

class Problem {
  final int id;
  final String title;
  final String description;
  final String difficulty;
  final List<String> topics;
  final List<StarterCode> starterCodes;
  final List<Example> examples;
  final List<Constraint> constraints;
  final List<Hint> hints;
  final List<TestCase> testcases;
  final int likes;

  Problem({
    required this.id,
    required this.title,
    required this.description,
    required this.difficulty,
    required this.topics,
    required this.starterCodes,
    required this.examples,
    required this.constraints,
    required this.hints,
    required this.testcases,
    required this.likes,
  });

  factory Problem.fromJson(Map<String, dynamic> json) {
    return Problem(
      id: json['id'] ?? 0,
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      difficulty: json['difficulty'] ?? '',
      topics:
          (json['topics'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
      starterCodes:
          (json['starterCodes'] as List<dynamic>?)
              ?.map((e) => StarterCode.fromJson(e))
              .toList() ??
          [],
      examples:
          (json['examples'] as List<dynamic>?)
              ?.map((e) => Example.fromJson(e))
              .toList() ??
          [],
      constraints:
          (json['constraints'] as List<dynamic>?)
              ?.map((e) => Constraint.fromJson(e))
              .toList() ??
          [],
      hints:
          (json['hints'] as List<dynamic>?)
              ?.map((e) => Hint.fromJson(e))
              .toList() ??
          [],
      testcases:
          (json['testcases'] as List<dynamic>?)
              ?.map((e) => TestCase.fromJson(e))
              .toList() ??
          [],
      likes: json['likes'] ?? 0,
    );
  }
}

File: leetlab/src/main/java/com/adviteey/leetlab/model/Problem.java
package com.adviteey.leetlab.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.Data;

import java.util.ArrayList;
import java.util.List;

@Data
@Entity
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
public class Problem {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String title;
    private String description;
    private long likes;
    @Enumerated(EnumType.STRING)
    private Difficulty difficulty;

    @ManyToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE})
    @JoinTable(
            name = "problem_topic",
            joinColumns = @JoinColumn(name = "problem_id"),
            inverseJoinColumns = @JoinColumn(name = "topic_id")
    )
    private List<Topic> topics = new ArrayList<>();

    @OneToMany(mappedBy = "problem", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Testcase> testcases = new ArrayList<>();

    @OneToMany(mappedBy = "problem", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<StarterCode>  starterCodes = new ArrayList<>();

    @OneToMany(mappedBy = "problem", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Example> examples = new ArrayList<>();

    @OneToMany(mappedBy = "problem", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<ProblemConstraint> problemConstraints = new ArrayList<>();

    @OneToMany(mappedBy = "problem", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Hint> hints = new ArrayList<>();

    @OneToMany(mappedBy = "problem", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Submission> submissions;
}